File: algorithmic-thinking/search-algorithms/bfs/bfs.js
/*
 * 广度优先搜索（BFS）- JavaScript 版本
 *
 * 使用队列按「层」遍历图或树：
 * - bfsTraversal: 遍历所有可达节点
 * - bfsShortestPath: 无权图最短路径
 *
 * 时间复杂度：O(V + E)
 * 空间复杂度：O(V)
 */

/**
 * BFS 遍历：返回从起点出发的访问顺序
 * @param {Object<string, string[]>} graph - 邻接表
 * @param {string} start - 起始节点
 */
function bfsTraversal(graph, start) {
  const visited = new Set();
  const queue = [start];
  const order = [];
  visited.add(start);

  while (queue.length > 0) {
    const node = queue.shift();
    order.push(node);
    for (const neighbor of graph[node] || []) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }
  return order;
}

/**
 * BFS 最短路径（无权图）
 * @param {Object<string, string[]>} graph
 * @param {string} start
 * @param {string} end
 * @returns {string[]|null} 最短路径（包含起点和终点），不存在返回 null
 */
function bfsShortestPath(graph, start, end) {
  if (start === end) return [start];
  const queue = [[start, [start]]];
  const visited = new Set([start]);

  while (queue.length > 0) {
    const [node, path] = queue.shift();
    for (const neighbor of graph[node] || []) {
      if (!visited.has(neighbor)) {
        if (neighbor === end) {
          return [...path, neighbor];
        }
        visited.add(neighbor);
        queue.push([neighbor, [...path, neighbor]]);
      }
    }
  }
  return null;
}

// 简单演示
if (require.main === module) {
  console.log("========== BFS (JavaScript) ==========");
  const graph = {
    A: ["B", "C"],
    B: ["A", "D", "E"],
    C: ["A", "F"],
    D: ["B"],
    E: ["B", "F"],
    F: ["C", "E"],
  };
  console.log("图:", graph);
  console.log("BFS 遍历从 A 开始 ->", bfsTraversal(graph, "A"));
  console.log("A 到 F 的最短路径 ->", bfsShortestPath(graph, "A", "F"));
}

module.exports = { bfsTraversal, bfsShortestPath };

File: algorithmic-thinking/search-algorithms/bfs/Bfs.java
package search.bfs;

import java.util.*;

/**
 * 广度优先搜索（BFS）- Java 版本
 *
 * - bfsTraversal: 从起点开始按层遍历所有可达节点
 * - bfsShortestPath: 在无权图中找到起点到终点的一条最短路径
 *
 * 时间复杂度：O(V + E)
 * 空间复杂度：O(V)
 */
public class Bfs {

    // BFS 遍历
    public static List<String> bfsTraversal(Map<String, List<String>> graph, String start) {
        List<String> order = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        Queue<String> queue = new ArrayDeque<>();

        visited.add(start);
        queue.offer(start);

        while (!queue.isEmpty()) {
            String node = queue.poll();
            order.add(node);
            for (String neighbor : graph.getOrDefault(node, Collections.emptyList())) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.offer(neighbor);
                }
            }
        }
        return order;
    }

    // BFS 最短路径（无权图）
    public static List<String> bfsShortestPath(Map<String, List<String>> graph, String start, String end) {
        if (start.equals(end)) {
            return Collections.singletonList(start);
        }
        Queue<List<String>> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        queue.offer(Collections.singletonList(start));
        visited.add(start);

        while (!queue.isEmpty()) {
            List<String> path = queue.poll();
            String node = path.get(path.size() - 1);

            for (String neighbor : graph.getOrDefault(node, Collections.emptyList())) {
                if (!visited.contains(neighbor)) {
                    List<String> newPath = new ArrayList<>(path);
                    newPath.add(neighbor);
                    if (neighbor.equals(end)) {
                        return newPath;
                    }
                    visited.add(neighbor);
                    queue.offer(newPath);
                }
            }
        }
        return null;
    }

    public static void main(String[] args) {
        System.out.println("========== BFS (Java) ==========");
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("A", Arrays.asList("B", "C"));
        graph.put("B", Arrays.asList("A", "D", "E"));
        graph.put("C", Arrays.asList("A", "F"));
        graph.put("D", Collections.singletonList("B"));
        graph.put("E", Arrays.asList("B", "F"));
        graph.put("F", Arrays.asList("C", "E"));

        System.out.println("图: " + graph);
        System.out.println("BFS 遍历从 A 开始 -> " + bfsTraversal(graph, "A"));
        System.out.println("A 到 F 的最短路径 -> " + bfsShortestPath(graph, "A", "F"));
    }
}

